design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+```
+The `CastEvaluator` class would be updated accordingly to generate the code that prevents null records from being written into the sink
+
+## Public APIS
+`CREATE STREAM/TABLE` with `NOT NULL` column:
+```sql
+CREATE stream pageviews(viewtime BIGINT, userid VARCHAR NOT NULL, pageid VARCHAR NOT NULL) ...
+```
+
+Given data in the pageviews topic:
+```sql
+3,,
+3,user,page
+2,,page1
+null,user,page
+```
+
+A `SELECT` query from the table would return only the two records that didn't violate the `NOT NULL` constraint:
+```sql
+ksql> select * from pageviews emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+|null           |user                                                            |page1                                                           |
+```
+
+Declaring a column as `NOT NULL` in a CSAS/CTAS would look like:
+```sql
+CREATE stream pageviews_no_null as select CAST(viewtime AS BIGINT NOT NULL), userid, pageid from pageviews;
+```
+The resulting stream's columns would all be `NOT NULL`
+
+```sql
+ksql> select * from pageviews_no_null emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+```
+
+Attempting to insert invalid records would error out:
+```sql
+ksql> INSERT INTO pageviews VALUES (5, NULL, 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "USERID"
+
+ksql> INSERT INTO pageviews VALUES (NULL, 'user3', 'page23'); # this insert succeeds
+
+ksql> INSERT INTO pageviews_no_null VALUES (NULL, 'user3', 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "VIEWTIME"
+```
+
+`NOT NULL` for arrays, maps, struct
+```sql
+CREATE STREAM valid_data (arr ARRAY<BIGINT NOT NULL>, mapping MAP<VARCHAR NOT NULL, BIGINT NOT NULL>, valid_struct STRUCT<c DOUBLE NOT NULL, d BOOLEAN>) ...
+```
+Records that have column values such as
```

Comment:
Just to make sure: We would drop the full record? (We won't only remove the corresponding array/map/struct entry?)